server/lib/src/idm/ldap.rs

```diff
@@ -525,6 +532,107 @@ impl LdapServer {
             },
         } // end match server op
     }
+
+    fn naming_context_from_base_dn(&self, _dn: &str) -> Result<&LdapNamingContext, OperationError> {
+        Ok(&self.default_naming_context)
+    }
+
+    async fn bind_target_from_bind_dn(
+        &self,
+        idms: &IdmServer,
+        dn: &str,
+        pw: &str,
+    ) -> Result<LdapBindTarget, OperationError> {
+        if dn.is_empty() {
+            if pw.is_empty() {
+                return Ok(LdapBindTarget::Account(UUID_ANONYMOUS));
+            } else {
+                // This is the path to access api-token logins.
+                return Ok(LdapBindTarget::ApiToken);
+            }
+        } else if dn == "dn=token" {
+            // Is the passed dn requesting token auth?
+            // We use dn= here since these are attr=value, and dn is a phantom so it will
+            // never be present or match a real value. We also make it an ava so that clients
+            // that over-zealously validate dn syntax are happy.
+            return Ok(LdapBindTarget::ApiToken);
+        }
+
+        if let Some(rdn) = self
+            .default_naming_context
+            .binddnre
+            .captures(dn)
+            .and_then(|caps| caps.name("val"))
+            .map(|v| v.as_str().to_string())
+        {
+            if rdn.is_empty() {
+                // That's weird ...
+                return Err(OperationError::NoMatchingEntries);
+            }
+
+            let mut idm_auth = idms.auth().await;
+            let uuid = idm_auth.qs_read.name_to_uuid(rdn.as_str()).map_err(|e| {
+                request_error!(err = ?e, ?rdn, "Error resolving rdn to target");
+                e
+            })?;
+
+            return Ok(LdapBindTarget::Account(uuid));
+        }
+
+        let basedn = &self.default_naming_context.basedn;
+        let app_binddnre = Regex::new(
+            format!("^(([^=,]+)=)(?P<usr>[^=,]+),(([^=,]+)=)(?P<app>[^=,]+)(,{basedn})$").as_str(),
+        )
+        .map_err(|_| OperationError::InvalidEntryState)?;
+
+        if let Some(caps) = app_binddnre.captures(dn) {
+            let app_rdn = caps
+                .name("app")
+                .map(|v| v.as_str().to_string())
+                .ok_or(OperationError::NoMatchingEntries)?;
+            let usr_rdn = caps
+                .name("usr")
+                .map(|v| v.as_str().to_string())
+                .ok_or(OperationError::NoMatchingEntries)?;
+            if app_rdn.is_empty() || usr_rdn.is_empty() {
+                return Err(OperationError::NoMatchingEntries);
+            }
+
+            let mut idms_prox_read = idms.proxy_read().await;
```

Comment:
This introduces a race condition. You should only have a single read txn over the whole operation. By having a read here that is dropped when the fn ends, then another when you call idms.auth(), there is a possibility that the application you are attempting to bind with that existed in the first transaction *no longer exists* when you get to the second. 

I think here at point you shouldn't be doing any db look ups at all - you should be only extracting the application name if one exists in the bind dn and that's it. It should be up to the call to application_auth_ldap to validate that name exists at that stage and then to do the bind.

Another way to think of this is that this part of the code is "preparing" all your needed inputs to build the auth event - the auth event is what makes all the choices and validations about if the application exists (or doesn't) and then validates the password. The auth event must be a fully and self contained call, without a need to rely on another part of the code to pre-vet anything.

Reply:
As another point, I think kanidm varies to other systems that we prefer to create and hold open a long-running read rather than open and closing a lot of smaller reads. The reason is that the copy-on-write structures we use are actually more effecient when you hold them for a longer time, and we want to assert from the start of any operation that "everything is isolated/atomic". If you ever open/close/open a txn in one request this breaks our ACID guarantees, and it actually introduces performance issues too.

Reply:
Ok, I have modified the code to delay application search. But doesn't `name_to_uuid()` call in `do_bind()` has the same problem?

Reply:
Yeah, I figured it out later. Thanks